include/pybind11/detail/internals.h

```diff
@@ -182,6 +185,16 @@ struct internals {
 #    endif // PYBIND11_INTERNALS_VERSION > 4
     // Unused if PYBIND11_SIMPLE_GIL_MANAGEMENT is defined:
     PyInterpreterState *istate = nullptr;
+
+#    if PYBIND11_INTERNALS_VERSION > 4
+    // Note that we have to use a std::string to allocate memory to ensure a unique address
+    // We want unique addresses since we use pointer equality to compare function records
+    std::string function_record_capsule_name = internals_function_record_capsule_name;
```

Comment:
> I am still not entirely sold on this since it could cause unexpected slowdowns when using passing callback builtins from other compilation units.

The `internals` object is reused between compilation units, even if using different pybind11 versions, via a capsule stored in builtins.

If the `internals` keys for retrieving that capsule don't match, using the `function_record` pointer is UB, for sure (unless people manipulate the macros for building `PYBIND11_INTERNALS_ID` in weird ways, but that's their own responsibility if they incur slowdowns then).

> We did test that this currently works to trigger the callback optimization right? It's just not skipping it everytime, right?

Yes, I inserted a print to check:

```
test_callbacks.py::test_cpp_function_roundtrip
LOOOK /usr/local/google/home/rwgk/forked/pybind11/include/pybind11/functional.h:70

LOOOK /usr/local/google/home/rwgk/forked/pybind11/include/pybind11/functional.h:70

LOOOK /usr/local/google/home/rwgk/forked/pybind11/include/pybind11/functional.h:70

LOOOK /usr/local/google/home/rwgk/forked/pybind11/include/pybind11/functional.h:70
PASSED
```

Reply:
> We did test that this currently works to trigger the callback optimization right? It's just not skipping it everytime, right?

That test is actually part of our existing test suite. 

https://github.com/pybind/pybind11/blob/14b7e335306907f2ebd2b84d635f035076689622/tests/test_callbacks.py#L93

Reply:
That makes me feel a bit better at least.